java/ql/src/semmle/code/java/frameworks/play/Play.qll

```diff
@@ -0,0 +1,155 @@
+  }
+}
+
+/**
+ * The type `play.libs.F.Promise<Result>`.
+ */
+class PlayAsyncResultPromise extends Member {
+  PlayAsyncResultPromise() {
+    exists(Class c |
+      c.hasQualifiedName("play.libs", "F") and
+      this = c.getAMember() and
+      this.getQualifiedName() = "F.Promise<Result>"
+    )
+  }
+}
+
+/**
+ * The type `java.util.concurrent.CompletionStage<Result>`.
+ */
+class PlayAsyncResultCompletionStage extends Type {
+  PlayAsyncResultCompletionStage() {
+    this.hasName("CompletionStage<Result>") and
+    this.getCompilationUnit().getPackage().hasName("java.util.concurrent")
+  }
+}
+
+/**
+ * The class `play.mvc.Controller` or a class that transitively extends it.
+ */
+class PlayController extends Class {
+  PlayController() {
+    this.extendsOrImplements*(any(Class t | t.hasQualifiedName("play.mvc", "Controller")))
+  }
+}
+
+/**
+ * A Play framework controller action method. These are mappings to route files.
+ *
+ * Sample Route - `POST  /login  @com.company.Application.login()`.
+ *
+ * Example - class gets `index` and `login` as valid action methods.
+ * ```java
+ * public class Application extends Controller {
+ *   public Result index(String username, String password) {
+ *     return ok("It works!");
+ *   }
+ *
+ *   public Result login() {
+ *     return ok("Log me In!");
+ *   }
+ * }
+ * ```
+ */
+class PlayControllerActionMethod extends Method {
+  PlayControllerActionMethod() {
+    this = any(PlayController c).getAMethod() and
+    (
+      this.getReturnType() instanceof PlayAsyncResultPromise or
+      this.getReturnType() instanceof PlayMvcResultClass or
+      this.getReturnType() instanceof PlayAsyncResultCompletionStage
+    )
+  }
+}
+
+/**
+ * A Play framework action method parameter. These are a source of user input.
+ *
+ * Example - `username` and `password` are action method parameters.
+ * ```java
+ * public class Application extends Controller {
+ *   public Result index(String username, String password) {
+ *     return ok("It works!");
+ *   }
+ * }
+ * ```
+ */
+class PlayActionMethodQueryParameter extends Parameter {
+  PlayActionMethodQueryParameter() {
+    exists(PlayControllerActionMethod a |
+      a.isPublic() and
+      this = a.getAParameter()
+    )
+  }
+}
+
+/**
+ * A PlayFramework HttpRequestHeader method, some of these are `headers`, `getQueryString`, `getHeader`.
+ */
+class PlayMvcHttpRequestHeaderMethods extends Method {
+  PlayMvcHttpRequestHeaderMethods() { this.getDeclaringType() instanceof PlayMvcHttpRequestHeader }
+
+  /**
+   * Gets a reference to the `getQueryString` method.
+   */
+  MethodAccess getAQueryStringAccess() {
+    this.hasName("getQueryString") and result = this.getAReference()
+  }
+}
+
+/**
+ * A PlayFramework results method, some of these are `ok`, `status`, `redirect`.
+ */
+class PlayMvcResultsMethods extends Method {
+  PlayMvcResultsMethods() { this.getDeclaringType() instanceof PlayMvcResultsClass }
+
```

Comment:
These two classes also don't appear used. Intentional or not?

Reply:
And these two?

Reply:
These are intentional, i will make a change in `FlowSources.qll` which is where `PlayMvcHttpRequestHeaderMethods` was meant to be used.